packages/teams-js/src/internal/videoUtils.ts

```diff
@@ -78,36 +108,142 @@ function createProcessedStreamGenerator(
+          // we need the timestamp to be unchanged from the oirginal frame, so we explicitly set it here.
+          timestamp: timestamp,
+        });
+        controller.enqueue(processedFrame);
+        videoFrame.close();
+        originalFrame.close();
+        (frameProcessedByApp as VideoFrame).close();
+      } catch (error) {
+        originalFrame.close();
+        this.notifyError(error);
+      }
+    } else {
+      this.notifyError('timestamp of the original video frame is null');
+    }
+  };
+
+  /**
+   * @hidden
+   * Extract video frame and metadata from the given texture.
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   */
+  private extractVideoFrameAndMetadata = async (
+    texture: VideoFrame,
+  ): Promise<{ videoFrame: VideoFrame; metadata: { audioInferenceResult?: Uint8Array } }> => {
+    if (inServerSideRenderingEnvironment()) {
+      throw errorNotSupportedOnPlatform;
+    }
+
+    if (texture.format !== 'NV12') {
+      this.notifyError('Unsupported video frame format');
+      throw new Error('Unsupported video frame format');
+    }
+
+    /**
+     * stream id for audio inference data
+     */
+    const AUDIO_INFERENCE_RESULT_STREAM_ID = 0x31646961;
+
+    // The rectangle of pixels to copy from the texture
+    const headerRect = { x: 0, y: 0, width: texture.codedWidth, height: 2 };
+    // it's in NV12 format, but the real data is in the y plane only.
+    const headerBuffer = new ArrayBuffer((headerRect.width * headerRect.height * 3) / 2);
+    await texture.copyTo(headerBuffer, { rect: headerRect });
+    const headerDataView = new Uint32Array(headerBuffer);
+    // const [ oneTextureId, version, frameRowOffset, frameFormat, frameWidth, frameHeight, multiStreamHeaderRowOffset, multiStreamCount ] = headerDataView;
+    const metadataRect = {
+      x: 0,
+      y: headerDataView[6], // multiStreamHeaderRowOffset
+      width: texture.codedWidth,
+      height: texture.codedHeight - headerDataView[6], // multiStreamHeaderRowOffset
+    };
+    const metadataBuffer = new ArrayBuffer((metadataRect.width * metadataRect.height * 3) / 2);
+    await texture.copyTo(metadataBuffer, { rect: metadataRect });
+    const metadata = new Uint32Array(metadataBuffer);
+    let audioInferenceResult: Uint8Array | undefined;
+    for (let i = 0, index = 0; i < headerDataView[7] /* multiStreamCount */; i++) {
+      const streamId = metadata[index++];
+      const streamDataOffset = metadata[index++];
+      const streamDataSize = metadata[index++];
+      const streamData = new Uint8Array(metadataBuffer, streamDataOffset, streamDataSize);
+      if (streamId === AUDIO_INFERENCE_RESULT_STREAM_ID) {
+        audioInferenceResult = streamData;
+      }
+    }
+
+    return {
+      videoFrame: new VideoFrame(texture as unknown as CanvasImageSource, {
+        timestamp: texture.timestamp,
+        visibleRect: {
+          x: 0,
+          y: headerDataView[2], // frameRowOffset
```

Comment:
3 more magic numbers

Reply:
added comments above to explain the memory layout